src/Evolution/DiscontinuousGalerkin/InboxTags.hpp

```diff
@@ -0,0 +1,126 @@
+ * is the TimeStepId beyond which the boundary terms are no longer valid, when
+ * using local time stepping.
+ *
+ * The TimeStepId is the neighboring element's next time step. When using local
+ * time stepping, the neighbor's boundary data is valid up until this time,
+ * which may include multiple local time steps. By receiving and storing the
+ * neighbor time step, the local element knows whether or not it should remove
+ * boundary data and expect new data to be sent from the neighbor.
+ *
+ * The ghost cell data (second element of the tuple) will be valid whenever
+ * DG-subcell scheme is being used. Whenever a DG-subcell scheme is being used,
+ * elements using DG and not FD/FV always send both the ghost cells and boundary
+ * correction data together. Elements using FD/FV send the ghost cells first
+ * followed by the boundary correction data once the element has received all
+ * neighbor ghost cell data. Note that the second send/receive only modifies the
+ * flux and the TimeStepId used for the flux validity range.
+ *
+ * When only a DG scheme (not a DG-subcell scheme) is used the ghost cell data
+ * will never be valid.
+ *
+ * In the DG-subcell scheme this tag is used both for communicating the ghost
+ * cell data needed for the FD/FV reconstruction step and the data needed for
+ * the boundary corrections. An element using a DG scheme will send both the
+ * ghost cells and the boundary correction terms simultaneously, while an
+ * element using FD/FV will first send the ghost cell data. The element
+ * receiving the ghost cells will store the data in this inbox tag and be able
+ * to do reconstruction once all neighboring ghost cells are received. In the
+ * mean time, as boundary correction data is received from neighbors it is added
+ * to this inbox tag. The inbox tag is not erased until after the boundary
+ * correction computation is complete. The reason for this design choice is that
+ * it is mostly the number of communications that dominates cost, not the size
+ * of the communication.
+ */
+template <size_t Dim>
+struct BoundaryCorrectionAndGhostCellsInbox {
+  using stored_type =
+      std::tuple<Mesh<Dim - 1>, std::optional<std::vector<double>>,
+                 std::optional<std::vector<double>>, ::TimeStepId>;
+
+ public:
+  using temporal_id = TimeStepId;
+  using type = std::map<
+      TimeStepId,
+      FixedHashMap<maximum_number_of_neighbors(Dim),
+                   std::pair<Direction<Dim>, ElementId<Dim>>, stored_type,
+                   boost::hash<std::pair<Direction<Dim>, ElementId<Dim>>>>>;
+
+  template <typename Inbox, typename ReceiveDataType>
+  static void insert_into_inbox(const gsl::not_null<Inbox*> inbox,
+                                const temporal_id& time_step_id,
+                                ReceiveDataType&& data) noexcept {
+    auto& current_inbox = (*inbox)[time_step_id];
+
+    if (auto it = current_inbox.find(data.first); it != current_inbox.end()) {
```

Comment:
c++ question: why are some uses of `data` wrapped in `std::forward<ReceiveDataType>` while others are not?

Reply:
The wrapped ones are ones where we are copying/moving the data out of the received data. The forward ensures that if the received data is an r-value reference that a move occurs for the element of the tuple. Does that make sense?